thread_pthread.c

```diff
@@ -101,6 +101,83 @@
 #  endif
 #endif
 
+static gvl_hook_t * rb_gvl_hooks = NULL;
+static pthread_rwlock_t rb_gvl_hooks_rw_lock = PTHREAD_RWLOCK_INITIALIZER;
+
+gvl_hook_t *
+rb_gvl_event_new(rb_gvl_callback callback, rb_event_flag_t event) {
+    gvl_hook_t *hook = ALLOC_N(gvl_hook_t, 1);
+    hook->callback = callback;
+    hook->event = event;
+
+    int r;
+    if ((r = pthread_rwlock_wrlock(&rb_gvl_hooks_rw_lock))) {
+        rb_bug_errno("pthread_rwlock_wrlock", r);
+    }
+
+    hook->next = rb_gvl_hooks;
+    ATOMIC_PTR_EXCHANGE(rb_gvl_hooks, hook);
```

Comment:
Is this atomic op needed, even with the rwlock?

Reply:
Yes. This is so that we can cheaply check wether at least one hook is registered without having to lock.

See how the various calls to `rb_gvl_execute_hooks` are under an unprotected `if (rb_gvl_hooks)`